### PR TITLE
[RISCV] Fix lowering of negative zero with Zdinx 32-bit

### PR DESCRIPTION
**Pull request by nemanjai**
The compiler currently abends with an impossible reg-to-reg copy when producing a negative zero FP immediate on RV32 with -Zdinx. This is because we emit a negation that uses FP registers. Emit the right node to produce correct code.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 1]
```diff
@@ -937,9 +938,14 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
       Res = CurDAG->getMachineNode(Opc, DL, VT, Imm);
 
     // For f64 -0.0, we need to insert a fneg.d idiom.
-    if (NegZeroF64)
-      Res = CurDAG->getMachineNode(RISCV::FSGNJN_D, DL, VT, SDValue(Res, 0),
+    if (NegZeroF64) {
+      if (Is64Bit)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:942]: Maybe flip Is64Bit and HasZdinx to remove the duplicate case for !Zdinx.

```
if (HasZdinx)
  Opc = Is64Bit ? RISCV::FSGNJN_D_INX : RISCV::FSGNJN_D_IN32X;
else
  Opc = RISCV::FSGNJN_D
```

or 

```
Opc = RISCV::FSGNJN_D;
if (HasZdinx)
  Opc = Is64Bit ? RISCV::FSGNJN_D_INX : RISCV::FSGNJN_D_IN32X;
```

Comment (nemanjai) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:942]: Sounds good, thank you.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by topperc** [comment]
> Glancing at the code post review, don't we have the same problem a couple of lines above for the FLI/ZFA codepath? Or is dinx and zfa an illegal combination in some way?

Zfa is defined to require F so it is not compatible with Zfinx/Zdinx. I think the thought was that several instructions in Zfa are redundant if you're using GPRs.

**Pull request merged**